src/control/server/server_utils.go

```diff
@@ -196,18 +183,46 @@ func prepBdevStorage(srv *server, iommuEnabled bool, hpiGetter common.GetHugePag
 	}
 
 	// Run prepare with reset first to release resources.
-	//
-	// TODO: should be passing root context into prepare request to
-	//       facilitate cancellation.
 	prepReq.Reset_ = true
 	if _, err := srv.ctlSvc.NvmePrepare(prepReq); err != nil {
 		srv.log.Errorf("automatic NVMe prepare reset failed: %s", err)
+	}
+
+	if hasBdevs {
+		// Use default value if bdevs have been configured but nr_hugepages unset in config.
+		if srv.cfg.NrHugepages < 0 {
+			srv.cfg.NrHugepages = 4096
+		}
+
+		// The config value is intended to be per-engine, so we need to allocate hugepages on
+		// each engine's numa node.
+		if srv.cfg.NrHugepages > 0 {
+			for _, ec := range srv.cfg.Engines {
+				nns = append(nns, fmt.Sprintf("%d", ec.GetNumaNode()))
+			}
+			// HugeNodeStr will allocate HugePageCount for each numa node in string
+			if len(nns) != 0 {
+				prepReq.HugeNodeStr = strings.Join(nns, ',')
+			}
+		} else {
+			srv.log.Errorf("bdevs detected in config but zero hugepages requested")
+		}
 	} else {
-		prepReq.Reset_ = false
-		if _, err := srv.ctlSvc.NvmePrepare(prepReq); err != nil {
-			srv.log.Errorf("automatic NVMe prepare failed: %s", err)
+		// If nr_hugepages unset and no bdevs in config, set minimum needed for scanning.
+		if srv.cfg.NrHugepages < 0 {
+			srv.cfg.NrHugepages = minHugePageCount
```

Comment:
This might not work, minHugepageCount is 128, this could make daos engine fail to start because it did not meet min requirement, could we assign 0 for this by default? and in the below try to assign preReq.HugePageCount = minHugePageCount, this is a bit tricky though.

Reply:
@NiuYawei @wangshilong apologies I submitted an incomplete patch set, please review the updated version (which has fixed the above)